src/main/ruby/truffleruby/core/truffle/polyglot.rb

```diff
@@ -34,6 +34,99 @@ def self.as_enumerable(object)
     Truffle::Interop.enumerable(object)
   end
 
+  module HasArrayElementsTrait
+
+    def size
+      Truffle::Interop.array_size(self)
+    end
+
+    alias_method :length, :size
+    alias_method :count, :size
+
+    def read_array_element(index)
+      Truffle::Interop.read_array_element(self, index)
+    end
+
+    def at(index)
+      index += length if index < 0
+      return nil if (index < 0 || index >= length)
+      read_array_element(index)
+    end
+
+    def first
+      read_array_element(0)
+    end
+
+    def last
+      read_array_element(length - 1)
+    end
+
+    def each(*args)
+      return enum_for(:each) unless block_given?
+      i = 0
+      while i < length
+        yield read_array_element(i)
+        i += 1
+      end
+    end
+
+    def each_with_index
+      return enum_for(:each) unless block_given?
+      i = 0
+      while i < length
+        yield read_array_element(i), i
+        i += 1
+      end
+    end
+
+  end
+
+  class ForeignObject
+    include HasArrayElementsTrait
+    include Enumerable
+
+    def has_array_elements?
+      Truffle::Interop.has_array_elements?(self)
+    end
+
+    def read_member(name)
+      Truffle::Interop.member(self, name)
+    end
+
+    def get_members(include_internal = false)
+      Truffle::Interop.members(self, include_internal)
+    end
```

Comment:
There should be no extra methods here, or at the very least they need to be `private`.